lib/block_view_message.go

```diff
@@ -825,95 +949,189 @@ func (bav *UtxoView) _connectMessagingGroup(
 		}
 
-		// Make sure the messagingMember public key and messaging key name are valid.
-		if err := ValidateGroupPublicKeyAndName(messagingMember.GroupMemberPublicKey[:], messagingMember.GroupMemberKeyName[:]); err != nil {
-			return 0, 0, nil, errors.Wrapf(err, "_connectMessagingGroup: "+
-				"Problem validating public key or messaging key for messagingMember (%v)", messagingMember.GroupMemberPublicKey[:])
+	case MessagingGroupOperationMuteMembers:
+		// Muting members assumes the group was already created.
+		if existingEntry == nil || existingEntry.isDeleted {
+			return 0, 0, nil, errors.Wrapf(RuleErrorMessagingGroupDoesntExist,
+				"_connectMessagingGroup: Can't mute members for a non-existing group")
+		}
+		// MUTING/UNMUTING functionality notes:
+		// In DeSo V3 Messages, Group Chat Owners can now mute or unmute members. This essentially acts like a
+		// "remove member from group" functionality, but can also be used to mute spammers in large channels.
+		// Note: A muted member can still cryptographically read the past AND future messages in the group, however,
+		// they cannot send messages to this group until they are unmuted by the group owner.
+		// Optimization Problem and Solution:
+		// Every time a new message arrives as a txn, we need to check inside _connectPrivateMessage() if the sender of
+		// the message is muted or not. This would decide whether we reject a message txn or not. However, to check
+		// that, we can't just fetch the entire MessagingGroupEntry which may contains 1000s if not 100,000s of members.
+		// Instead, we will make usage of the membership index. We will especially see this in the flushing logic.
+		for _, existingMutedMember := range existingEntry.MuteList {
+			newMuteList = append(newMuteList, existingMutedMember)
 		}
 
-		// Now make sure messagingMember's MessagingGroupKey has already been added to UtxoView or DB.
-		// We encrypt the groupMessagingKey to recipients' messaging keys.
-		memberMessagingGroupKey := NewMessagingGroupKey(
-			messagingMember.GroupMemberPublicKey, messagingMember.GroupMemberKeyName[:])
-		memberGroupEntry := bav.GetMessagingGroupKeyToMessagingGroupEntryMapping(memberMessagingGroupKey)
-		// The messaging key has to exist and cannot be deleted.
-		if memberGroupEntry == nil || memberGroupEntry.isDeleted {
-			return 0, 0, nil, errors.Wrapf(
-				RuleErrorMessagingMemberKeyDoesntExist, "_connectMessagingGroup: "+
-					"Problem verifying messaing key for messagingMember (%v)", messagingMember.GroupMemberPublicKey[:])
+		for _, newlyMutedMember := range txMeta.MessagingGroupMembers {
+			// Make sure GroupOwner is not muting herself
+			if reflect.DeepEqual(newlyMutedMember.GroupMemberPublicKey[:], existingEntry.GroupOwnerPublicKey[:]) {
+				return 0, 0, nil, errors.Wrapf(RuleErrorMessagingGroupOwnerMutingSelf,
+					"_connectMessagingGroup: GroupOwner cannot mute herself (%v).", existingEntry.GroupOwnerPublicKey[:])
+			}
```

Comment:
should we prevent a user from muting themselves? also can any member mute any other member or is it only a group owner? This is a potential avenue for abuse in a groupchat. If I don't like  you, I just mute you. I think it's okay if that's the situation, but want to make sure we're considering the UX of this feature.

Reply:
Only the owner can mute others. During the connect logic for the messaging group txn we fetch an entry based on the txn.PublicKey, so you can only modify your own groups.